userspace/libscap/scap.h

```diff
@@ -719,6 +745,23 @@ scap_threadinfo* scap_get_proc_table(scap_t* handle);
 */
 int32_t scap_get_stats(scap_t* handle, OUT scap_stats* stats);
 
+/*!
+  \brief Get engine statistics (including counters and `bpftool prog show` like stats)
+
+  \param handle Handle to the capture instance.
+  \param flags holding statistics category flags.
+  \param nstats Pointer reflecting number of statistics in returned buffer.
+  \param rc Pointer to return code.
+
+  \return Pointer to a \ref scap_stats_v2 structure filled with the statistics.
+*/
+struct scap_stats_v2* scap_get_stats_v2(scap_t* handle, uint32_t flags, OUT uint32_t* nstats, OUT int32_t* rc);
```

Comment:
Thanks for addressing the suggestion. So right now we create a new stats buffer every time `scap_get_stats_v2` is invoked, having a alloc + free flow at every iteration. Likely (at least for Falco), we'd need an additional memory copy of the returned struct to push it in the asynchronous output framework queue in order to push it into a Falco alert.

Now, why don't we just allocate this buffer once per engine instead? Each engine knows beforehand what the max amount of stats it can return can be. Key benefits:
- We would save the cost of a malloc + free at every invocation
- We would reduce memory copied by one, since engines could potentially just write their counters right in the stats struct if they can (they would be total owners of that piece of memory after all)
- We would not need the `scap_free_stats_v2` symbol at all, lifetime of the buffer will be the same of the engine itself

I'm not super strong on this, but please let me know your thinking. @gnosek @Andreagit97

Reply:
Sounds more elegant to me that way. Let's do it right would say.

Now even wondering about the CPU and memory metrics we added in libsinsp, given the libbpf metrics additions here took an unexpected amazing turn of complete stats refactor I even wonder if we should move them all into scap in here as well (not in this PR of course)? Because the allocation and free would be the same problem with the resource utilization additions. It's now so generic and versatile that we really should do that, it would make the Falco consumer logic so simple for the new metrics logs.

Reply:
I think we could avoid pushing extra logic into libscap. However, if the flexibility of the new stats C definition is what we really need we can just use it to represent sinsp stats as well.

Reply:
> libscap end users may want to benefit from those resource utilization metrics as well as they really are not sinsp specific?

I look at libscap the other way: is $THING something that absolutely has to live in scap and nowhere else? If not, don't put it there. Realistically, you probably won't use raw libscap but only sinsp anyway.

But of course, we should make the sinsp stats use this interface if we can (if we can't, we picked the wrong interface :|).